crates/bls12381/src/fields/fp.rs

```diff
@@ -0,0 +1,445 @@
+
+    pub fn alloc_is_zero<CS>(&self, cs: &mut CS) -> Result<AllocatedBit, SynthesisError>
+    where
+        CS: ConstraintSystem<F>,
+    {
+        self.0.alloc_is_zero(&mut cs.namespace(|| "val =? 0"))
+    }
+
+    pub fn add<CS>(&self, cs: &mut CS, value: &Self) -> Result<Self, SynthesisError>
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let res = self.0.add(&mut cs.namespace(|| "a + b"), &value.0)?;
+        Ok(Self(res))
+    }
+
+    pub fn sub<CS>(&self, cs: &mut CS, value: &Self) -> Result<Self, SynthesisError>
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let res = self.0.sub(&mut cs.namespace(|| "a - b"), &value.0)?;
+        Ok(Self(res))
+    }
+
+    pub fn mul<CS>(&self, cs: &mut CS, value: &Self) -> Result<Self, SynthesisError>
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let res = self.0.mul(&mut cs.namespace(|| "a * b"), &value.0)?;
+        Ok(Self(res))
+    }
+
+    pub fn inverse<CS>(&self, cs: &mut CS) -> Result<Self, SynthesisError>
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let res = self.0.inverse(&mut cs.namespace(|| "x.inverse()"))?;
+        Ok(Self(res))
+    }
+
+    pub fn div_unchecked<CS>(&self, cs: &mut CS, value: &Self) -> Result<Self, SynthesisError>
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let res = self.0.divide(&mut cs.namespace(|| "a div b"), &value.0)?;
+        Ok(Self(res))
+    }
+
+    pub fn square<CS>(&self, cs: &mut CS) -> Result<Self, SynthesisError>
+    where
+        CS: ConstraintSystem<F>,
+    {
+        self.mul(&mut cs.namespace(|| "x.square()"), self)
+    }
+
+    pub fn double<CS>(&self, cs: &mut CS) -> Result<Self, SynthesisError>
+    where
+        CS: ConstraintSystem<F>,
+    {
+        self.add(&mut cs.namespace(|| "x.double()"), self)
+    }
+
+    pub fn mul_const<CS>(&self, cs: &mut CS, value: &BigInt) -> Result<Self, SynthesisError>
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let res = self
+            .0
+            .mul_const(&mut cs.namespace(|| "a * constval"), value)?;
+        Ok(Self(res))
+    }
+
+    pub fn neg<CS>(&self, cs: &mut CS) -> Result<Self, SynthesisError>
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let res = self.0.neg(&mut cs.namespace(|| "-a"))?;
+        Ok(Self(res))
+    }
+
+    pub fn conditionally_select<CS>(
+        cs: &mut CS,
+        z0: &Self,
+        z1: &Self,
+        condition: &Boolean,
+    ) -> Result<Self, SynthesisError>
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let res = Bls12381Fp::<F>::conditionally_select(
+            &mut cs.namespace(|| "cond val"),
+            &z0.0,
+            &z1.0,
+            condition,
+        )?;
+        Ok(Self(res))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use bellpepper_core::test_cs::TestConstraintSystem;
+    use pasta_curves::Fp;
+
+    use expect_test::{expect, Expect};
+    fn expect_eq(computed: usize, expected: Expect) {
```

Comment:
Agreed on all counts, we can improve this later, I just wanted to get `expect-test` in this initial PR because every test has the constraint count and there's quite a few of those. I think using `proptest` would be a neat followup